File: examples and misc/Cpp/polymorphism.cpp
#include <iostream>
using namespace std;

// Base class
class Animal {
public:
    virtual void speak() {
        cout << "Animal makes a sound." << endl;
    }
};

// Derived class
class Dog : public Animal {
public:
    void speak() override {
        cout << "Dog barks." << endl;
    }
};

// Derived class
class Cat : public Animal {
public:
    void speak() override {
        cout << "Cat meows." << endl;
    }
};

int main() {
    Animal* animals[] = {new Dog(), new Cat()};
    for (Animal* animal : animals) {
        animal->speak();
    }
    return 0;
}

File: examples and misc/Java/Polymorphism.java
public class Polymorphism {
    // Base class
    static class Animal {
        void speak() {
            System.out.println("Animal makes a sound.");
        }
    }

    // Derived class
    static class Dog extends Animal {
        @Override
        void speak() {
            System.out.println("Dog barks.");
        }
    }

    // Derived class
    static class Cat extends Animal {
        @Override
        void speak() {
            System.out.println("Cat meows.");
        }
    }

    static class Bird extends Animal {
        @Override
        void speak() {
            System.out.println("Bird chirps.");
        }
    }
    public static void main(String[] args) {
        Animal[] animals = {new Dog(), new Cat(), new Bird()};
        for (Animal animal : animals) {
            animal.speak();
        }
    }
}
